### PR TITLE
fix: datepicker doesn't close on first selection of date on month change

### PR DESCRIPTION
**Pull request by SaiCharanChetpelly31**
Fixes issue #30701 

Added cypress test for the fix

https://github.com/appsmithorg/appsmith/assets/119920490/d8e92377-6d27-49f9-8839-e18b8b6aae65

## Summary by CodeRabbit

- **New Features**
	- Added new tests to the DatePicker widget ensuring functionality and reliability.
- **Bug Fixes**
	- Improved date selection logic to handle null values and formatted dates correctly.

### CODE REVIEW

**Comment by SaiCharanChetpelly31** [comment]
Hi @ApekshaBhosale Can you please review this PR?

**Comment by SaiCharanChetpelly31** [comment]
@ApekshaBhosale   Gentle Reminder: Could you please review this PR?

**Comment by SaiCharanChetpelly31** [comment]
Hi @ApekshaBhosale  ,

I hope you're well. I’m curious and excited to hear your thoughts on my pull request raised a while ago. Looking forward to your review!
Thank you.

File: app/client/src/widgets/DatePickerWidget2/component/index.tsx [hunk 1]
```diff
@@ -481,6 +481,18 @@ class DatePickerComponent extends React.Component<
...
+    if (selectedDate === null) {
+      this.setState({ selectedDate: "" });
+      this.props.onDateSelected("");
+    } else {
+      const formattedDate = moment(selectedDate).format(ISO_DATE_FORMAT);
+      this.setState({ selectedDate: formattedDate }, () => {
+        this.props.onDateSelected(formattedDate);
+      });
+    }
+  };
```
[the remark this answers is not in the document]
Comment (ApekshaBhosale) [reply app/client/src/widgets/DatePickerWidget2/component/index.tsx:495]: @SaiCharanChetpelly31 please consider this to make change to  your code

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/Datepicker/DatePicker4_spec.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        .first()
+        .find(".DayPicker-Day")
+        .then(($days) => {
+          const randomIndex = Math.floor(Math.random() * $days.length);
+          cy.wrap($days[randomIndex]).click();
+        });
+    }
+
+    it("1.should close datepicker on single click when month is changed", function () {
+      cy.openPropertyPane("datepickerwidget2");
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/Datepicker/DatePicker4_spec.js:32]: use locators from WigetLocator file

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/Datepicker/DatePicker4_spec.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      cy.get(".t--widget-datepickerwidget2 input").click();
+      cy.get(".DayPicker").should("be.visible");
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/Datepicker/DatePicker4_spec.js:34]: use `AssertElementVisibility` helper

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/Datepicker/DatePicker4_spec.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+      navigateAndSelectRandomDate("prev");
+      cy.get(".t--widget-datepickerwidget2 input").click();
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/Datepicker/DatePicker4_spec.js:37]: use `GetNClick`  helper

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/Datepicker/DatePicker4_spec.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      navigateAndSelectRandomDate("prev");
+      cy.get(".t--widget-datepickerwidget2 input").click();
```
Comment (ApekshaBhosale) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/Datepicker/DatePicker4_spec.js:39]: use `GetNClick` helper

**Review by ApekshaBhosale — CHANGES_REQUESTED** [review]

**Comment by SaiCharanChetpelly31** [comment]
@ApekshaBhosale I accidentally deleted my forked repository, but I have raised another PR with the suggested changes addressing the comments. Could you please review [this PR](https://github.com/appsmithorg/appsmith/pull/34679)?

**Pull request closed without merging**